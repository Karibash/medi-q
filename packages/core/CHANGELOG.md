# @medi-q/core

## 0.3.0

### Minor Changes

- [#10](https://github.com/Karibash/medi-q/pull/10) [`1334ae0`](https://github.com/Karibash/medi-q/commit/1334ae08f4a68b06ce5a7f33ec0cc12a4426e504) Thanks [@Karibash](https://github.com/Karibash)! - Restrict the units that can be used to px, em, and rem

### Patch Changes

- [#10](https://github.com/Karibash/medi-q/pull/10) [`b237028`](https://github.com/Karibash/medi-q/commit/b23702862c87a2762c95d41edf709ea607d7dc05) Thanks [@Karibash](https://github.com/Karibash)! - Make the initial value of the unit conversion to em

## 0.0.3

### Patch Changes

- [`8fcbce8`](https://github.com/Karibash/medi-q/commit/8fcbce84f79bdaa051c78fee6db657b92e4decc9) Thanks [@Karibash](https://github.com/Karibash)! - Update README.md

## 0.0.2

### Patch Changes

- [`1ec0be4`](https://github.com/Karibash/medi-q/commit/1ec0be4c7c6aef361142c89582fd5554b2f5a511) Thanks [@Karibash](https://github.com/Karibash)! - Fix an error when normalizing breakpoints

## 0.0.1

### Major Changes

- First release.
