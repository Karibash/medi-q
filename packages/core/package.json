{
  "name": "@medi-q/core",
  "version": "1.0.0",
  "description": "Type-safe media queries.",
  "main": "dist/medi-q-core.cjs.js",
  "module": "dist/medi-q-core.esm.js",
  "unpkg": "dist/medi-q-core.umd.min.js",
  "jsdelivr": "dist/medi-q-core.umd.min.js",
  "umd:main": "dist/medi-q-core.umd.min.js",
  "types": "dist/medi-q-core.cjs.d.ts",
  "browser": {
    "./dist/medi-q-core.cjs.js": "./dist/medi-q-core.browser.cjs.js",
    "./dist/medi-q-core.esm.js": "./dist/medi-q-core.browser.esm.js"
  },
  "files": [
    "src",
    "dist",
    "LICENSE",
    "README.md"
  ],
  "keywords": [
    "typescript",
    "media-query",
    "responsive",
    "breakpoint"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Karibash/medi-q.git"
  },
  "author": "Karibash",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/Karibash/medi-q/issues"
  },
  "homepage": "https://github.com/Karibash/medi-q/tree/main/packages/core",
  "dependencies": {
    "@karibash/pixel-units": "^2.3.0"
  },
  "devDependencies": {
    "typescript": "^4.4.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "preconstruct": {
    "umdName": "mediq"
  }
}
