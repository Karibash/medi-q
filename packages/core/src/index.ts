import { ConvertLengthUnitArgsOptions, LengthUnitSuffix } from '@karibash/pixel-units';

import { normalizedBreakPoints, queryInputToQueryString } from './internals';
import { BreakPoints, MediQ } from './types';

export * from './types';

export const createMediQ = (
  breakPoints: BreakPoints,
  unit: Extract<LengthUnitSuffix, 'px' | 'em' | 'rem'> = 'em',
  options?: Pick<ConvertLengthUnitArgsOptions, 'em' | 'rem'>,
): MediQ => {
  const keys = Object.keys(breakPoints);
  const _breakPoints = normalizedBreakPoints(breakPoints, unit, options);
  return (queryInput) => queryInputToQueryString(queryInput, _breakPoints, keys);
};
