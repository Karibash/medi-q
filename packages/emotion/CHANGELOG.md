# @medi-q/emotion

## 0.0.1
### Major Changes

- First release.
