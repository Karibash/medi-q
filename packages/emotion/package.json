{
  "name": "@medi-q/emotion",
  "version": "0.0.3",
  "description": "medi-q for emotion.",
  "main": "dist/medi-q-emotion.cjs.js",
  "module": "dist/medi-q-emotion.esm.js",
  "unpkg": "dist/medi-q-emotion.umd.min.js",
  "jsdelivr": "dist/medi-q-emotion.umd.min.js",
  "umd:main": "dist/medi-q-emotion.umd.min.js",
  "types": "dist/medi-q-emotion.cjs.d.ts",
  "browser": {
    "./dist/medi-q-emotion.cjs.js": "./dist/medi-q-emotion.browser.cjs.js",
    "./dist/medi-q-emotion.esm.js": "./dist/medi-q-emotion.browser.esm.js"
  },
  "files": [
    "src",
    "dist",
    "LICENSE",
    "README.md"
  ],
  "keywords": [
    "typescript",
    "react",
    "emotion",
    "media-query",
    "responsive",
    "breakpoint"
  ],
  "repository": {
    "type": "git",
    "url": "git+https://github.com/Karibash/medi-q.git"
  },
  "author": "Karibash",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/Karibash/medi-q/issues"
  },
  "homepage": "https://github.com/Karibash/medi-q/tree/main/packages/emotion",
  "dependencies": {
    "@babel/runtime": "^7.15.4",
    "@medi-q/core": "^0.0.3",
    "@medi-q/react": "^0.0.3"
  },
  "peerDependencies": {
    "@emotion/react": "^11.0.0-rc.0",
    "react": ">=16.8.0"
  },
  "peerDependenciesMeta": {
    "@emotion/react": {
      "optional": true
    },
    "react": {
      "optional": true
    }
  },
  "devDependencies": {
    "@emotion/react": "^11.4.1",
    "react": "^17.0.2",
    "typescript": "^4.4.3"
  },
  "publishConfig": {
    "access": "public"
  },
  "preconstruct": {
    "umdName": "mediqEmotion"
  }
}
