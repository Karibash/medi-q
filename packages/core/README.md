# @medi-q/core

The @medi-q/core package is a framework agnostic for simply generating media queries.

## 🚀 Installation

```
$ npm install @medi-q/core
```

## 👏 Getting Started

By using the createMediQ function, you can create a helper function to generate media queries.
You can also generate media queries with multiple conditions by connecting the queries with and or.

```tsx
import { createMediQ } from '@medi-q/core';

const mediQ = createMediQ({
  tiny: '400px',
  small: '600px',
  medium: '800px',
  large: '1000px',
});

const maxSmall = mediQ('max-small');
const minMedium = mediQ('min-medium');
const minSmallAndMaxMedium = mediQ('min-small-and-max-medium');

const onResize = (): void => {
  if (matchMedia(maxSmall).matches) {
    console.log('isLessThanSmall');
  } else if (matchMedia(minMedium).matches) {
    console.log('isGreaterThanMedium');
  } else if (matchMedia(minSmallAndMaxMedium).matches) {
    console.log('isBetweenSmallAndMedium');
  }
};
onResize();
window.onresize = onResize;
```
