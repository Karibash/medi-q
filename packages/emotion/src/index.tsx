import React, { ReactNode } from 'react';
import { MediQ, MediQInput } from '@medi-q/core';
import { MediQProvider } from '@medi-q/react';
import { ThemeProvider as EmotionThemeProvider, Theme } from '@emotion/react';

declare module '@emotion/react' {
  export interface Theme {
    mediQ: MediQ;
  }
}

export type ThemeProviderProps = {
  theme: Omit<Theme, 'mediQ'>;
  mediQ: MediQ;
  children: ReactNode;
};

export const ThemeProvider: React.FC<ThemeProviderProps> = ({
  mediQ,
  theme,
  children,
}) => {
  return (
    <MediQProvider mediQ={mediQ}>
      <EmotionThemeProvider theme={{ ...theme, mediQ: (input: MediQInput) => `@media ${mediQ(input)}` }}>
        {children}
      </EmotionThemeProvider>
    </MediQProvider>
  );
};
