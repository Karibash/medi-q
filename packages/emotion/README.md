# @medi-q/emotion

This is a package for using medi-q with [emotion](https://github.com/emotion-js/emotion).

## 🚀 Installation

```
$ npm install @medi-q/core @medi-q/react @medi-q/emotion
```

## 👏 Getting Started

Use the ThemeProvider component to pass the Theme object and MediQ object to the subordinate components.

```tsx
import React from 'react';
import { BreakPoints, createMediQ } from '@medi-q/core';
import { ThemeProvider } from '@medi-q/emotion';

import { theme } from './theme';

const breakPoints: BreakPoints = {
  tiny: '400px',
  small: '600px',
  medium: '800px',
  large: '1000px',
};

const App: React.FC = () => {
  return (
    <ThemeProvider theme={theme} mediQ={createMediQ(breakPoints)}>
      ...
    </ThemeProvider>
  );
};

export default App;
```

By using the useMediQ hook in the lower-level component, you can execute a media query.
You can also generate media queries with multiple conditions by connecting the queries with and or.

```tsx
import React from 'react';
import styled from '@emotion/styled';
import { useMediQ } from '@medi-q/react';

const Wrapper = styled.div`
  background: ${props => props.theme.background};

  ${props => props.theme.mediQ('max-medium')} {
    background: blue;
  }
`;

const Page: React.FC = () => {
  const isLessThanSmall = useMediQ('max-small');
  const isGreaterThanMedium = useMediQ('min-medium');
  const isBetweenSmallAndMedium = useMediQ('min-small-and-max-medium');
  return (
    <Wrapper>
      {isLessThanSmall && <div>isLessThanSmall</div>}
      {isGreaterThanMedium && <div>isGreaterThanMedium</div>}
      {isBetweenSmallAndMedium && <div>isBetweenSmallAndMedium</div>}
    </Wrapper>
  );
};

export default Page;
```
