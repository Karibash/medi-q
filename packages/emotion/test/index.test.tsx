/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render } from '@testing-library/react';
import styled from '@emotion/styled';
import { createSerializer } from '@emotion/jest';

import { createMediQ } from '@medi-q/core';
import { ThemeProvider } from '@medi-q/emotion';

expect.addSnapshotSerializer(createSerializer());

document.documentElement.style.fontSize = '16px';
const mediQ = createMediQ({
  tiny: '400px',
  small: '600px',
  medium: '800px',
  large: '1000px',
});

describe('@medi-q/styled', () => {
  it('invalid query prefix', () => {
    expect(() => {
      const mock = jest.spyOn(console, 'error').mockImplementation(jest.fn());
      const Component = styled.div`
        ${props => {
          // @ts-ignore
          return props.theme.mediQ('invalid-tiny');
        }} {
          color: #000000;
        }
      `;

      render(
        <ThemeProvider theme={{}} mediQ={mediQ}>
          <Component />
        </ThemeProvider>,
      );
      expect(mock).toBeCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });

  it('invalid query breakpoint', () => {
    expect(() => {
      const mock = jest.spyOn(console, 'error').mockImplementation(jest.fn());
      const Component = styled.div`
        ${props => {
          // @ts-ignore
          return props.theme.mediQ('min-invalid');
        }} {
          color: #000000;
        }
      `;

      render(
        <ThemeProvider theme={{}} mediQ={mediQ}>
          <Component />
        </ThemeProvider>,
      );
      expect(mock).toBeCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });

  it('invalid query conjunctions', () => {
    expect(() => {
      const mock = jest.spyOn(console, 'error').mockImplementation(jest.fn());
      const Component = styled.div`
        ${props => {
          // @ts-ignore
          return props.theme.mediQ('min-tiny-invalid-max-small');
        }} {
          color: #000000;
        }
      `;

      render(
        <ThemeProvider theme={{}} mediQ={mediQ}>
          <Component />
        </ThemeProvider>,
      );
      expect(mock).toBeCalledTimes(1);
    }).toThrowErrorMatchingSnapshot();
  });

  it('min query', () => {
    const Component = styled.div`
      color: #000000;

      ${props => props.theme.mediQ('min-tiny')} {
        background: #000001;
      }

      ${props => props.theme.mediQ('min-small')} {
        background: #000002;
      }

      ${props => props.theme.mediQ('min-medium')} {
        background: #000003;
      }

      ${props => props.theme.mediQ('min-large')} {
        background: #000004;
      }
    `;

    const { container } = render(
      <ThemeProvider theme={{}} mediQ={mediQ}>
        <Component />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('max query', () => {
    const Component = styled.div`
      color: #000000;

      ${props => props.theme.mediQ('max-tiny')} {
        background: #000001;
      }

      ${props => props.theme.mediQ('max-small')} {
        background: #000002;
      }

      ${props => props.theme.mediQ('max-medium')} {
        background: #000003;
      }

      ${props => props.theme.mediQ('max-large')} {
        background: #000004;
      }
    `;

    const { container } = render(
      <ThemeProvider theme={{}} mediQ={mediQ}>
        <Component />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('composite query', () => {
    const Component = styled.div`
      color: #000000;

      ${props => props.theme.mediQ('min-tiny-and-max-tiny')} {
        background: #000001;
      }

      ${props => props.theme.mediQ('min-small-and-max-small')} {
        background: #000002;
      }

      ${props => props.theme.mediQ('min-medium-and-max-medium')} {
        background: #000003;
      }

      ${props => props.theme.mediQ('min-large-and-max-large')} {
        background: #000004;
      }

      ${props => props.theme.mediQ('max-tiny-or-min-tiny')} {
        background: #000005;
      }

      ${props => props.theme.mediQ('max-small-or-min-small')} {
        background: #000006;
      }

      ${props => props.theme.mediQ('max-medium-or-min-medium')} {
        background: #000007;
      }

      ${props => props.theme.mediQ('max-large-or-min-large')} {
        background: #000008;
      }
    `;

    const { container } = render(
      <ThemeProvider theme={{}} mediQ={mediQ}>
        <Component />
      </ThemeProvider>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
