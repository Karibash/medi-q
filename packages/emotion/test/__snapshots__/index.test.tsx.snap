// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@medi-q/styled composite query 1`] = `
.emotion-0 {
  color: #000000;
}

@media (min-width: 25em) and (max-width: 25em) {
  .emotion-0 {
    background: #000001;
  }
}

@media (min-width: 37.5em) and (max-width: 37.5em) {
  .emotion-0 {
    background: #000002;
  }
}

@media (min-width: 50em) and (max-width: 50em) {
  .emotion-0 {
    background: #000003;
  }
}

@media (min-width: 62.5em) and (max-width: 62.5em) {
  .emotion-0 {
    background: #000004;
  }
}

@media (max-width: 25em), (min-width: 25em) {
  .emotion-0 {
    background: #000005;
  }
}

@media (max-width: 37.5em), (min-width: 37.5em) {
  .emotion-0 {
    background: #000006;
  }
}

@media (max-width: 50em), (min-width: 50em) {
  .emotion-0 {
    background: #000007;
  }
}

@media (max-width: 62.5em), (min-width: 62.5em) {
  .emotion-0 {
    background: #000008;
  }
}

<div
  class="emotion-0"
/>
`;

exports[`@medi-q/styled invalid query breakpoint 1`] = `"Invalid query. Please use the values shown on the right for breakpoints: [tiny,small,medium,large]."`;

exports[`@medi-q/styled invalid query conjunctions 1`] = `"Invalid query. Please use \\"and\\" or \\"or\\" for conjunctions."`;

exports[`@medi-q/styled invalid query prefix 1`] = `"Invalid query. Please use \\"max\\" or \\"min\\" as the query prefix."`;

exports[`@medi-q/styled max query 1`] = `
.emotion-0 {
  color: #000000;
}

@media (max-width: 25em) {
  .emotion-0 {
    background: #000001;
  }
}

@media (max-width: 37.5em) {
  .emotion-0 {
    background: #000002;
  }
}

@media (max-width: 50em) {
  .emotion-0 {
    background: #000003;
  }
}

@media (max-width: 62.5em) {
  .emotion-0 {
    background: #000004;
  }
}

<div
  class="emotion-0"
/>
`;

exports[`@medi-q/styled min query 1`] = `
.emotion-0 {
  color: #000000;
}

@media (min-width: 25em) {
  .emotion-0 {
    background: #000001;
  }
}

@media (min-width: 37.5em) {
  .emotion-0 {
    background: #000002;
  }
}

@media (min-width: 50em) {
  .emotion-0 {
    background: #000003;
  }
}

@media (min-width: 62.5em) {
  .emotion-0 {
    background: #000004;
  }
}

<div
  class="emotion-0"
/>
`;
